### PR TITLE
Collapse statistics in normal explain plan

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/4144

# Rationale for this change

It is almost impossible to get any useful information out of `explain analyze` for a parquet exec when it has more than 1-2 files. See https://github.com/apache/arrow-datafusion/issues/4144 for examples

# What changes are included in this PR?

Collapses all the statisics from parquet exec with the same name when showing in explain plan

Before:
```
|                   |           ParquetExec: limit=None, partitions=[Users/alamb/Software/arrow-datafusion2/parquet-testing/data/alltypes_plain.parquet], predicate=timestamp_col_max@0 > 1233446400000000000, projection=[id, float_col, timestamp_col], metrics=[output_rows=8, elapsed_compute=1ns, spill_count=0, spilled_bytes=0, mem_used=0, bytes_scanned{filename=Users/alamb/Software/arrow-datafusion2/parquet-testing/data/alltypes_plain.parquet}=259, pushdown_rows_filtered{filename=Users/alamb/Software/arrow-datafusion2/parquet-testing/data/alltypes_plain.parquet}=0, row_groups_pruned{filename=Users/alamb/Software/arrow-datafusion2/parquet-testing/data/alltypes_plain.parquet}=0, predicate_evaluation_errors{filename=Users/alamb/Software/arrow-datafusion2/parquet-testing/data/alltypes_plain.parquet}=0, page_index_rows_filtered{filename=Users/alamb/Software/arrow-datafusion2/parquet-testing/data/alltypes_plain.parquet}=0, num_predicate_creation_errors=0, page_index_eval_time{filename=Users/alamb/Software/arrow-datafusion2/parquet-testing/data/alltypes_plain.parquet}=2ns, time_elapsed_opening=558.733µs, time_elapsed_processing=1.103953ms, pushdown_eval_time{filename=Users/alamb/Software/arrow-datafusion2/parquet-testing/data/alltypes_plain.parquet}=2ns, time_elapsed_scanning=656.927µs] |
```

After (still not great but easier easier):
```
|                   |           ParquetExec: limit=None, partitions=[Users/alamb/Software/arrow-datafusion2/parquet-testing/data/alltypes_plain.parquet], predicate=timestamp_col_max@0 > 1233446400000000000, projection=[id, float_col, timestamp_col], metrics=[output_rows=8, elapsed_compute=1ns, spill_count=0, spilled_bytes=0, mem_used=0, predicate_evaluation_errors=0, pushdown_rows_filtered=0, bytes_scanned=259, num_predicate_creation_errors=0, row_groups_pruned=0, page_index_rows_filtered=0, time_elapsed_opening=588.861µs, time_elapsed_scanning=644.732µs, time_elapsed_processing=1.115933ms, page_index_eval_time=2ns, pushdown_eval_time=2ns] |
```

# Are these changes tested?
Yes
# Are there any user-facing changes?
Yes

### CODE REVIEW

File: datafusion/core/src/physical_plan/metrics/mod.rs [hunk 1]
```diff
@@ -259,27 +259,23 @@ impl MetricsSet {
     }
 
     /// Returns returns a new derived `MetricsSet` where all metrics
-    /// that had the same name and partition=`Some(..)` have been
+    /// that had the same name have been
     /// aggregated together. The resulting `MetricsSet` has all
     /// metrics with `Partition=None`
-    pub fn aggregate_by_partition(&self) -> Self {
+    pub fn aggregate_by_name(&self) -> Self {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/metrics/mod.rs:265]: This is the core change

File: datafusion/core/src/physical_plan/metrics/mod.rs [hunk 2]
```diff
@@ -140,7 +140,7 @@ impl Metric {
     }
 
     /// Add a new label to this metric
-    pub fn with(mut self, label: Label) -> Self {
+    pub fn with_label(mut self, label: Label) -> Self {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/metrics/mod.rs:143]: Drive by cleanup to make the setter conform to standard `with_` naming

**Review by Ted-Jiang — APPROVED** [review]
@alamb Thanks for doing this, this metric system is so subtle.❤️

**Review by andygrove — APPROVED** [review]

**Comment by alamb** [comment]
> @alamb Thanks for doing this, this metric system is so subtle.❤️

Yeah, I didn't really know what I was doing when I originally implemented it. It is definitely due for a revamp if someone has the time. I suspect it could be made much simpler

**Pull request merged**